src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -376,6 +376,226 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
+//                 int length, V v, M m,
+//                 UnaryMaskedOperation<V, M> defaultImpl) {
+//
+// public static
+// <V, M>
+// V ternaryMaskedOp(int oprId, Class<? extends V> vmClass, Class<? extends M> maskClass, Class<?> elementType,
+//                   int length, V v1, V v2, V v3, M m,
+//                   TernaryMaskedOperation<V, M> defaultImpl) {
+//
+bool LibraryCallKit::inline_vector_nary_masked_operation(int n) {
+  const TypeInt*     opr          = gvn().type(argument(0))->isa_int();
+  const TypeInstPtr* vector_klass = gvn().type(argument(1))->isa_instptr();
+  const TypeInstPtr* mask_klass   = gvn().type(argument(2))->isa_instptr();
+  const TypeInstPtr* elem_klass   = gvn().type(argument(3))->isa_instptr();
+  const TypeInt*     vlen         = gvn().type(argument(4))->isa_int();
+
+  if (opr == NULL || vector_klass == NULL || mask_klass == NULL || elem_klass == NULL || vlen == NULL ||
+      !opr->is_con() || vector_klass->const_oop() == NULL || mask_klass->const_oop() == NULL ||
+      elem_klass->const_oop() == NULL || !vlen->is_con()) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** missing constant: opr=%s vclass=%s maskclass=%s etype=%s vlen=%s",
+                    NodeClassNames[argument(0)->Opcode()],
+                    NodeClassNames[argument(1)->Opcode()],
+                    NodeClassNames[argument(2)->Opcode()],
+                    NodeClassNames[argument(3)->Opcode()],
+                    NodeClassNames[argument(4)->Opcode()]);
+    }
+    return false; // not enough info for intrinsification
+  }
+  ciType* elem_type = elem_klass->const_oop()->as_instance()->java_mirror_type();
+  if (!elem_type->is_primitive_type()) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not a primitive bt=%d", elem_type->basic_type());
+    }
+    return false; // should be primitive type
+  }
+  if (!is_klass_initialized(vector_klass)) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** klass argument not initialized");
+    }
+    return false;
+  }
+
+  BasicType elem_bt = elem_type->basic_type();
+  int num_elem = vlen->get_con();
+  int opc = VectorSupport::vop2ideal(opr->get_con(), elem_bt);
+  int sopc = VectorNode::opcode(opc, elem_bt);
+  if ((opc != Op_CallLeafVector) && (sopc == 0)) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** operation not supported: opc=%s bt=%s", NodeClassNames[opc], type2name(elem_bt));
+    }
+    return false; // operation not supported
+  }
+  ciKlass* vbox_klass = vector_klass->const_oop()->as_instance()->java_lang_Class_klass();
+  const TypeInstPtr* vbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, vbox_klass);
+
+  // "argument(n + 5)" should be the mask object. We assume it is "null" when no mask
+  // is used to control this operation.
+  bool is_masked_op = gvn().type(argument(n + 5)) != TypePtr::NULL_PTR;
+  if (is_vector_mask(vbox_klass)) {
+    assert(!is_masked_op, "mask operations do not need mask to control");
```

Comment:
Another invariant to check here is mask input has to be either "always null" or "never null".

Currently, it will be checked when mask is unboxed, but it makes sense to make the check evident.